C error - calling a function in an or block in a for loop

### Describe the bug

The compiler outputs a C error.

### Reproduction Steps

```v
fn option() ![]int {
        return []
}

fn func() {}

fn main() {
        for _ in option() or {
                func()
                return
        } {
        }
}
```

### Expected Behavior

No C error.

### Current Behavior

```
❯ v test.v
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/test.01JK0X4FX3D5AZKAP61X7K7RB2.tmp.c:6600: error: ';' expected (got "return")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
```

### Possible Solution

_No response_

### Additional Information/Context

(happy to have made vreduce, I did not want to reduce by hand the 2.5k lines of code to try to reproduce this)

### V version

V 0.4.9 b8c7892

### Environment details (OS name and version, etc.)

Current V version: V 0.4.9 b8c7892, timestamp: 2025-02-01 13:41:38 +0200
V self compiling (-prod -o v2)...
V built successfully as executable "v".
|V full version      |V 0.4.9 6ac1d08.b8c7892
|:-------------------|:-------------------
|OS                  |linux, "Void Linux"
|Processor           |4 cpus, 64bit, little endian, Intel(R) Pentium(R) Gold 7505 @ 2.00GHz
|Memory              |16.83GB/23.16GB
|                    |
|V executable        |/home/nopana/v/v
|V last modified time|2025-02-01 14:01:50
|                    |
|V home dir          |OK, value: /home/nopana/v
|VMODULES            |OK, value: /home/nopana/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/nopana/projects/vreduce
|                    |
|Git version         |git version 2.47.1
|V git status        |weekly.2025.03-68-gb8c7892f
|.git/config present |true
|                    |
|cc version          |cc (GCC) 13.2.0
|gcc version         |gcc (GCC) 13.2.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.39
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2131,9 +2131,10 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 				}
 			}
 		} else {
-			if i > 0 {
+			if i > 0 && stmt is ast.Return {
 				last_stmt := stmts[i - 1]
-				if last_stmt is ast.ExprStmt && last_stmt.expr is ast.CallExpr {
+				if last_stmt is ast.ExprStmt && last_stmt.expr is ast.CallExpr
+					&& !g.out.last_n(2).contains(';') {
 					g.writeln(';')
 				}
 			}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2131,6 +2131,12 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 				}
 			}
 		} else {
+			if i > 0 {
+				last_stmt := stmts[i - 1]
+				if last_stmt is ast.ExprStmt && last_stmt.expr is ast.CallExpr {
+					g.writeln(';')
+				}
+			}
 			g.stmt(stmt)
 			if (g.inside_if_option || g.inside_if_result || g.inside_match_option
 				|| g.inside_match_result) && stmt is ast.ExprStmt {
@@ -4012,10 +4018,11 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	mut sum_type_dot := '.'
 	mut field_typ := ast.void_type
 	mut is_option_unwrap := false
+	is_iface_or_sumtype := sym.kind in [.interface, .sum_type]
 	if f := g.table.find_field_with_embeds(sym, node.field_name) {
 		field_sym := g.table.sym(f.typ)
 		field_typ = f.typ
-		if sym.kind in [.interface, .sum_type] {
+		if is_iface_or_sumtype {
 			g.write('(*(')
 		}
 		is_option := field_typ.has_flag(.option)
@@ -4036,7 +4043,11 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 					for i, typ in field.smartcasts {
 						if i == 0 && (is_option_unwrap || nested_unwrap) {
 							deref := if g.inside_selector {
-								'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+								if is_iface_or_sumtype {
+									'*'.repeat(field.smartcasts.last().nr_muls())
+								} else {
+									'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+								}
 							} else if sym.kind == .interface && !typ.is_ptr()
 								&& field.orig_type.has_flag(.option) {
 								''
@@ -4138,7 +4149,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			return
 		}
 	} else {
-		if sym.kind in [.interface, .sum_type] {
+		if is_iface_or_sumtype {
 			g.write('(*(')
 		}
 	}

--- a/vlib/v/tests/options/option_for_or_block_test.v
+++ b/vlib/v/tests/options/option_for_or_block_test.v
@@ -1,0 +1,14 @@
+fn option() ![]int {
+	return []
+}
+
+fn func() {}
+
+fn test_main() {
+	for _ in option() or {
+		func()
+		return
+	} {
+	}
+	assert true
+}

